backend/ring/ring.go

```diff
@@ -238,29 +230,44 @@ func (r *Ring) Next(ctx context.Context) (string, error) {
 				// This seems to happen from time to time, but is very hard
 				// to reproduce.
 				// TODO: find some way of avoiding this.
-				delChan <- struct{}{}
+				ch <- struct{}{}
 			}
 		}
 	}()
+	return ch
+}
+
+// Next returns the next owned item in the Ring and advances the iteration. If
+// the Ring does not contain any items owned by this client, then Next will
+// block until the context is cancelled. If the Ring contains no items
+// whatsoever, then ErrEmptyRing will be returned.
+func (r *Ring) Next(ctx context.Context) (string, error) {
+	watchCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	wc := r.client.Watch(watchCtx, r.Name, clientv3.WithPrefix())
+	if wc == nil {
+		return "", ctx.Err()
+	}
+	wait := notifyDeletes(watchCtx, wc)
 	for {
-		select {
-		case <-ctx.Done():
-			return "", ctx.Err()
-		default:
+		if err := ctx.Err(); err != nil {
```

Comment:
I'm probably missing something (the _context_ pkg is hard haha) but shouldn't it be `watchCtx.Err()`?

Reply:
It doesn't matter too much in this case. Contexts are hierarchical, and the cancellation we are looking for comes from outside the function. watchCtx will inherit any cancellation from ctx, aside from its own CancelFunc.

However, now that you mention it, it seems a bit inconsistent to create a new context and then not use it here, so I think I will change it to watchCtx.

Reply:
I updated the PR to use a single context.